server/gameconnection.py

```diff
@@ -311,19 +322,26 @@ async def handle_operation_complete(self, army, secondary, delta):
             )
             row = await result.fetchone()
             if not row:
-                self._logger.debug("can't find coop map: %s", self.game.map_file_path)
+                self._logger.debug(
+                    "can't find coop map: %s", self.game.map_file_path
+                )
                 return
             mission = row["id"]
 
-            await conn.execute(
-                coop_leaderboard.insert().values(
-                    mission=mission,
-                    gameuid=self.game.id,
-                    secondary=secondary,
-                    time=delta,
-                    player_count=len(self.game.players),
+            # Each player in a co-op game will send the OperationComplete
+            # message but we only need to perform this insert once
+            already_saved = getattr(self.game, '_coop_leaderboard_saved', False)
```

Comment:
IMO, just make this an instance variable on `CoopGame`. Which brings up a point: what if we see an `OperationComplete` message on a game that is not a `CoopGame`? I think that either means there's a bug and the server didn't use the right game class, or the game is doing something really weird probably due to mods. I think either way though, probably in that case no leaderboard entry should be created and a warning should be logged.

Reply:
Isn't that what `if self.game.validity != ValidityState.COOP_NOT_RANKED:` is checking?

Reply:
Hm, I guess that would do it yea. Although it's more of a happy accident :P.